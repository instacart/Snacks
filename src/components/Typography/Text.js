import React from 'react'
import Radium from 'radium'
import PropTypes from 'prop-types'
import { colors } from '../../styles'

// Not using `margin` or `padding` shorthand  to avoid any future
// Radium warnings with combining shorthand/longhand properties.
const baseStyles = {
  paddingTop: 0,
  paddingRight: 0,
  paddingBottom: 0,
  paddingLeft: 0,
  marginTop: 0,
  marginRight: 0,
  marginBottom: 0,
  marginLeft: 0,
  color: colors.GRAY_13
}

// Matches the Open Sans font weights
// https://fonts.google.com/specimen/Open+Sans?selection.family=Open+Sans
const fontWeights = {
  light: 300,
  regular: 400,
  semiBold: 600,
  bold: 700
}

const variantStyles = {
  'T.92': {
    fontSize: '92px',
    lineHeight: '138px',
    letterSpacing: '0',
    fontWeight: fontWeights.regular
  },
  'T.82': {
    fontSize: '82px',
    lineHeight: '123px',
    letterSpacing: '0',
    fontWeight: fontWeights.regular
  },
  'T.72': {
    fontSize: '72px',
    lineHeight: '108px',
    letterSpacing: '0',
    fontWeight: fontWeights.regular
  },
  'T.64': {
    fontSize: '64px',
    lineHeight: '96px',
    letterSpacing: '0',
    fontWeight: fontWeights.regular
  },
  'T.58': {
    fontSize: '58px',
    lineHeight: '87px',
    letterSpacing: '0',
    fontWeight: fontWeights.regular
  },
  'T.46': {
    fontSize: '46px',
    lineHeight: '69px',
    letterSpacing: '0',
    fontWeight: fontWeights.regular
  },
  'T.36': {
    fontSize: '36px',
    lineHeight: '54px',
    letterSpacing: '0.03em',
    fontWeight: fontWeights.regular
  },
  'T.28': {
    fontSize: '28px',
    lineHeight: '42px',
    letterSpacing: '0.05em',
    fontWeight: fontWeights.bold
  },
  'T.22': {
    fontSize: '22px',
    lineHeight: '32px',
    letterSpacing: '0.03em',
    fontWeight: fontWeights.regular
  },
  'T.18': {
    fontSize: '18px',
    lineHeight: '32px',
    letterSpacing: '0.03em',
    fontWeight: fontWeights.regular
  },
  'T.16': {
    fontSize: '16px',
    lineHeight: '24px',
    letterSpacing: '0.03em',
    fontWeight: fontWeights.regular
  },
  'T.14': {
    fontSize: '14px',
    lineHeight: '21px',
    letterSpacing: '0.03em',
    fontWeight: fontWeights.regular
  },
  'T.12': {
    fontSize: '12px',
    lineHeight: '18px',
    letterSpacing: '0.05em',
    fontWeight: fontWeights.regular
  },
  'T.11': {
    fontSize: '11px',
    lineHeight: '16px',
    letterSpacing: '0.05em',
    fontWeight: fontWeights.regular
  }
}

const htmlTagMapping = {
  'T.92': 'h1',
  'T.82': 'h1',
  'T.72': 'h1',
  'T.64': 'h1',
  'T.58': 'h1',
  'T.46': 'h1',
  'T.36': 'h2',
  'T.28': 'h2',
  'T.22': 'h2',
  'T.18': 'p',
  'T.16': 'p',
  'T.14': 'p',
  'T.12': 'p',
  'T.11': 'p'
}

function Text({ variant, style, children, elementType, fontWeight, ...restProps }) {
  const ElementType = elementType || htmlTagMapping[variant]
  const finalStyles = {
    ...baseStyles,
    ...variantStyles[variant],
    ...(fontWeight && { fontWeight: fontWeights[fontWeight] }),
    ...style
  }

  return (
    <ElementType style={finalStyles} {...restProps}>
      {children}
    </ElementType>
  )
}

Text.propTypes = {
  /** Typography variant to be used. */
  variant: PropTypes.oneOf(Object.keys(variantStyles)).isRequired,

  /** Text content to be rendered. */
  children: PropTypes.node.isRequired,

  /** Overrides the default HTML element type. Each typography variant has a default HTML element (e.g. h1, p),
   * but this is only a best guess. Always ensure you're using the correct header markup -- element
   * types with Snacks typography should be used to describe document structure, not visual style. */
  elementType: PropTypes.string,

  /** Overrides the variant's default font weight. */
  fontWeight: PropTypes.oneOf(Object.keys(fontWeights)),

  /** Any custom inline styles. Useful for things like gutters and responsive styles. */
  style: PropTypes.shape({})
}

export default Radium(Text)
