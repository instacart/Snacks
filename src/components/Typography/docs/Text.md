Consistent typography and hierarchy is our most useful tool in creating a clear and understandable product for our customers.

Provide a `variant` prop to specify which typography variant to use:

```jsx static
<Text variant="T.14">
  T.14 (body text)
</Text>
```

With an optional `fontWeight`:

```jsx static
<Text variant="T.14" fontWeight="bold">
  T.14 (bold body text)
</Text>
```

With an optional `elementType`. `Text` comes with a default typography variant => HTML element mapping, but you may need to pass a custom element type to ensure your markup is correct:

```jsx static
<Text variant="T.18" elementType="h2">
  T.18 (title/subtitle text)
</Text>
```

Provide an optional style object (will override any existing styles as it's applied last). Use only when other component props are insufficient (e.g. for gutters):

```jsx static
<Text variant="T.14" style={{ marginBottom: 10 }}>
  T.14 (body text)
</Text>
```

```jsx noeditor
const TypographyTableRow = require('./helpers').TypographyTableRow

<div style={{marginTop: '24px'}}>
  <style dangerouslySetInnerHTML={{__html: `
    .typography-table td {
      padding: 10px;
      vertical-align: top;
    }
    .typography-table-variant {
      white-space: nowrap;
    }
  `}} />
  <hr />
  <div style={{marginTop: '24px'}}>
    <Text variant="T.28" fontWeight="bold" style={{marginBottom: 16}}>Type Guidelines: Web and Mobile Web</Text>
    <Text variant="T.16">Guidelines on font treatment along with suggested use cases</Text>
  </div>
  <table className="typography-table" style={{width: '100%', marginTop: '24px'}}>
    <tbody>
      <TypographyTableRow
        variant="T.92"
        usage={['Web: Oversized screen titles', 'Mobile: Calling out metrics and numbers on dashboards or other relevant screens']}
        example="Snacks"
      />
      <TypographyTableRow
        variant="T.82"
        usage={['Web: Oversized screen titles', 'Mobile: Calling out metrics and numbers on dashboards or other relevant screens']}
        example="Snacks"
      />
      <TypographyTableRow
        variant="T.72"
        usage={['Web: Oversized screen titles', 'Mobile: Calling out metrics and numbers on dashboards or other relevant screens']}
        example="Snacks"
      />
      <TypographyTableRow
        variant="T.64"
        usage={['Web: Oversized screen titles', 'Mobile: Calling out metrics and numbers on dashboards or other relevant screens']}
        example="Snacks"
      />
      <TypographyTableRow
        variant="T.58"
        usage={['Web: Title size', 'Mobile: Calling out metrics and numbers on dashboards or other relevant screens']}
        example="Snacks"
      />
      <TypographyTableRow
        variant="T.46"
        usage={['Web: Top level headers. Ideal as titles on web or subtitles paired with larger sizes']}
      />
      <TypographyTableRow
        variant="T.36"
        usage={['Web: Standard size for web titles', 'Mobile: Large titles']}
      />
      <TypographyTableRow
        variant="T.28"
        usage={['Web: Titles for sections, modals, trays, etc.', 'Mobile: Large titles']}
      />
      <TypographyTableRow
        variant="T.22"
        usage={['Web: Subtitles and section titles', 'Mobile: Large titles']}
      />
      <TypographyTableRow
        variant="T.18"
        usage={['Web: Subtitles and section titles', 'Mobile: Title. Also used as body size on shopper products']}
      />
      <TypographyTableRow
        variant="T.16"
        usage={['Web: Body text where 14 is too small', 'Mobile: Title, subtitle, or body text']}
      />
      <TypographyTableRow
        variant="T.14"
        usage={['Web: Body text, usually good for paragraphs', 'Mobile: Subtitle or body text']}
      />
      <TypographyTableRow
        variant="T.12"
        usage={['Web: Subtext for de-emphaiszed information', 'Mobile: Body text (e.g. on item cards). More commonly seen on customers side']}
      />
      <TypographyTableRow
        variant="T.11"
        usage={['Web: DO NOTE USE! Does not meet accessibility standards', 'Mobile: Subtext for de-emphasized information like legal jargon. More commonly seen on customers side']}
      />
    </tbody>
  </table>
</div>
```
