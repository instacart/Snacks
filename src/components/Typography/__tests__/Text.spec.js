import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import Text from '../Text'

const getComponent = props => <Text {...props}>Don't talk about snacks</Text>

describe('Text', () => {
  it('renders all typography variants correctly', () => {
    const variants = [
      'T.92',
      'T.82',
      'T.72',
      'T.64',
      'T.58',
      'T.46',
      'T.36',
      'T.28',
      'T.22',
      'T.18',
      'T.16',
      'T.14',
      'T.12',
      'T.11',
    ]

    variants.forEach(variantName => {
      const tree = renderer.create(getComponent({ variant: variantName })).toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  it('applies custom inline styles', () => {
    const component = getComponent({ variant: 'T.14', style: { marginBottom: '10px' } })
    const tree = renderer.create(component).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('applies a custom html element type', () => {
    const component = getComponent({ elementType: 'span', variant: 'T.14' })
    const tree = renderer.create(component).toJSON()

    expect(tree.type).toEqual('span')
  })

  it('applies a custom font weight', () => {
    const component = getComponent({ variant: 'T.28', fontWeight: 'semiBold' })
    const style = window.getComputedStyle(mount(component).getDOMNode())
    expect(style.fontWeight).toEqual('600')
  })
})
