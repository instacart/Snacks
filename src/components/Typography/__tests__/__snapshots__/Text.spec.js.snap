// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text applies custom inline styles 1`] = `
.emotion-0 {
  padding-top: 0;
  padding-right: 0;
  padding-bottom: 0;
  padding-left: 0;
  margin-top: 0;
  margin-right: 0;
  margin-bottom: 10px;
  margin-left: 0;
  color: #212121;
  font-size: 14px;
  line-height: 21px;
  -webkit-letter-spacing: 0.03em;
  -moz-letter-spacing: 0.03em;
  -ms-letter-spacing: 0.03em;
  letter-spacing: 0.03em;
  font-weight: 400;
}

<p
  className="emotion-0"
>
  Don't talk about snacks
</p>
`;

exports[`Text renders all typography variants correctly 1`] = `
.emotion-0 {
  padding-top: 0;
  padding-right: 0;
  padding-bottom: 0;
  padding-left: 0;
  margin-top: 0;
  margin-right: 0;
  margin-bottom: 0;
  margin-left: 0;
  color: #212121;
  font-size: 92px;
  line-height: 138px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  font-weight: 400;
}

<h1
  className="emotion-0"
>
  Don't talk about snacks
</h1>
`;

exports[`Text renders all typography variants correctly 2`] = `
.emotion-0 {
  padding-top: 0;
  padding-right: 0;
  padding-bottom: 0;
  padding-left: 0;
  margin-top: 0;
  margin-right: 0;
  margin-bottom: 0;
  margin-left: 0;
  color: #212121;
  font-size: 82px;
  line-height: 123px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  font-weight: 400;
}

<h1
  className="emotion-0"
>
  Don't talk about snacks
</h1>
`;

exports[`Text renders all typography variants correctly 3`] = `
.emotion-0 {
  padding-top: 0;
  padding-right: 0;
  padding-bottom: 0;
  padding-left: 0;
  margin-top: 0;
  margin-right: 0;
  margin-bottom: 0;
  margin-left: 0;
  color: #212121;
  font-size: 72px;
  line-height: 108px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  font-weight: 400;
}

<h1
  className="emotion-0"
>
  Don't talk about snacks
</h1>
`;

exports[`Text renders all typography variants correctly 4`] = `
.emotion-0 {
  padding-top: 0;
  padding-right: 0;
  padding-bottom: 0;
  padding-left: 0;
  margin-top: 0;
  margin-right: 0;
  margin-bottom: 0;
  margin-left: 0;
  color: #212121;
  font-size: 64px;
  line-height: 96px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  font-weight: 400;
}

<h1
  className="emotion-0"
>
  Don't talk about snacks
</h1>
`;

exports[`Text renders all typography variants correctly 5`] = `
.emotion-0 {
  padding-top: 0;
  padding-right: 0;
  padding-bottom: 0;
  padding-left: 0;
  margin-top: 0;
  margin-right: 0;
  margin-bottom: 0;
  margin-left: 0;
  color: #212121;
  font-size: 58px;
  line-height: 87px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  font-weight: 400;
}

<h1
  className="emotion-0"
>
  Don't talk about snacks
</h1>
`;

exports[`Text renders all typography variants correctly 6`] = `
.emotion-0 {
  padding-top: 0;
  padding-right: 0;
  padding-bottom: 0;
  padding-left: 0;
  margin-top: 0;
  margin-right: 0;
  margin-bottom: 0;
  margin-left: 0;
  color: #212121;
  font-size: 46px;
  line-height: 69px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  font-weight: 400;
}

<h1
  className="emotion-0"
>
  Don't talk about snacks
</h1>
`;

exports[`Text renders all typography variants correctly 7`] = `
.emotion-0 {
  padding-top: 0;
  padding-right: 0;
  padding-bottom: 0;
  padding-left: 0;
  margin-top: 0;
  margin-right: 0;
  margin-bottom: 0;
  margin-left: 0;
  color: #212121;
  font-size: 36px;
  line-height: 54px;
  -webkit-letter-spacing: 0.03em;
  -moz-letter-spacing: 0.03em;
  -ms-letter-spacing: 0.03em;
  letter-spacing: 0.03em;
  font-weight: 400;
}

<h2
  className="emotion-0"
>
  Don't talk about snacks
</h2>
`;

exports[`Text renders all typography variants correctly 8`] = `
.emotion-0 {
  padding-top: 0;
  padding-right: 0;
  padding-bottom: 0;
  padding-left: 0;
  margin-top: 0;
  margin-right: 0;
  margin-bottom: 0;
  margin-left: 0;
  color: #212121;
  font-size: 28px;
  line-height: 42px;
  -webkit-letter-spacing: 0.05em;
  -moz-letter-spacing: 0.05em;
  -ms-letter-spacing: 0.05em;
  letter-spacing: 0.05em;
  font-weight: 700;
}

<h2
  className="emotion-0"
>
  Don't talk about snacks
</h2>
`;

exports[`Text renders all typography variants correctly 9`] = `
.emotion-0 {
  padding-top: 0;
  padding-right: 0;
  padding-bottom: 0;
  padding-left: 0;
  margin-top: 0;
  margin-right: 0;
  margin-bottom: 0;
  margin-left: 0;
  color: #212121;
  font-size: 22px;
  line-height: 32px;
  -webkit-letter-spacing: 0.03em;
  -moz-letter-spacing: 0.03em;
  -ms-letter-spacing: 0.03em;
  letter-spacing: 0.03em;
  font-weight: 400;
}

<h2
  className="emotion-0"
>
  Don't talk about snacks
</h2>
`;

exports[`Text renders all typography variants correctly 10`] = `
.emotion-0 {
  padding-top: 0;
  padding-right: 0;
  padding-bottom: 0;
  padding-left: 0;
  margin-top: 0;
  margin-right: 0;
  margin-bottom: 0;
  margin-left: 0;
  color: #212121;
  font-size: 18px;
  line-height: 32px;
  -webkit-letter-spacing: 0.03em;
  -moz-letter-spacing: 0.03em;
  -ms-letter-spacing: 0.03em;
  letter-spacing: 0.03em;
  font-weight: 400;
}

<p
  className="emotion-0"
>
  Don't talk about snacks
</p>
`;

exports[`Text renders all typography variants correctly 11`] = `
.emotion-0 {
  padding-top: 0;
  padding-right: 0;
  padding-bottom: 0;
  padding-left: 0;
  margin-top: 0;
  margin-right: 0;
  margin-bottom: 0;
  margin-left: 0;
  color: #212121;
  font-size: 16px;
  line-height: 24px;
  -webkit-letter-spacing: 0.03em;
  -moz-letter-spacing: 0.03em;
  -ms-letter-spacing: 0.03em;
  letter-spacing: 0.03em;
  font-weight: 400;
}

<p
  className="emotion-0"
>
  Don't talk about snacks
</p>
`;

exports[`Text renders all typography variants correctly 12`] = `
.emotion-0 {
  padding-top: 0;
  padding-right: 0;
  padding-bottom: 0;
  padding-left: 0;
  margin-top: 0;
  margin-right: 0;
  margin-bottom: 0;
  margin-left: 0;
  color: #212121;
  font-size: 14px;
  line-height: 21px;
  -webkit-letter-spacing: 0.03em;
  -moz-letter-spacing: 0.03em;
  -ms-letter-spacing: 0.03em;
  letter-spacing: 0.03em;
  font-weight: 400;
}

<p
  className="emotion-0"
>
  Don't talk about snacks
</p>
`;

exports[`Text renders all typography variants correctly 13`] = `
.emotion-0 {
  padding-top: 0;
  padding-right: 0;
  padding-bottom: 0;
  padding-left: 0;
  margin-top: 0;
  margin-right: 0;
  margin-bottom: 0;
  margin-left: 0;
  color: #212121;
  font-size: 12px;
  line-height: 18px;
  -webkit-letter-spacing: 0.05em;
  -moz-letter-spacing: 0.05em;
  -ms-letter-spacing: 0.05em;
  letter-spacing: 0.05em;
  font-weight: 400;
}

<p
  className="emotion-0"
>
  Don't talk about snacks
</p>
`;

exports[`Text renders all typography variants correctly 14`] = `
.emotion-0 {
  padding-top: 0;
  padding-right: 0;
  padding-bottom: 0;
  padding-left: 0;
  margin-top: 0;
  margin-right: 0;
  margin-bottom: 0;
  margin-left: 0;
  color: #212121;
  font-size: 11px;
  line-height: 16px;
  -webkit-letter-spacing: 0.05em;
  -moz-letter-spacing: 0.05em;
  -ms-letter-spacing: 0.05em;
  letter-spacing: 0.05em;
  font-weight: 400;
}

<p
  className="emotion-0"
>
  Don't talk about snacks
</p>
`;
